feature/BJS2-97260-post-view-analytics

Deathroels opened the pull request.

Deathroels committed:
-add kafka impl

Deathroels committed:
-add KafkaProducerConfig

Deathroels committed:
-add PostViewEvent

Deathroels committed:
-add PostViewPublisher, need to fix publish method

Deathroels committed:
-try to add events in facede?

Deathroels committed:
-add kafka bootstrap-servers in app.yaml
-add impl kafka starter

Deathroels committed:
-add names for Bean's
-change bean duplicate problem

Deathroels committed:
-moved kafka topic name in application.yaml

Deathroels committed:
-impl spring-boot-starter-aop

Deathroels committed:
-moved kafka topic name to application.yaml

Deathroels committed:
-add aspect and PublishPostEvent, need todo async and mode parameters in aspect

Deathroels committed:
-todo mark, need to fix annotation
-add topic in infra service
-todo analytics service side

Deathroels committed:
-remove wrong import

Deathroels committed:
BJS2-97260:
-add KafkaProducerConfig

Deathroels committed:
BJS2-97260:
-add PublishPostEvent interface

Deathroels committed:
BJS2-97260:
-add PostEventAspect

Deathroels committed:
BJS2-97260:
-remove mode from aspect, need to delete async param, to default true

Deathroels committed:
BJS2-97260:
-change in KafkaTemplate, to <String, Object>

Deathroels committed:
change application.yaml kafka publishers

Deathroels committed:
change annotation for PostEventAspect

Deathroels committed:
deleted default param for custom annotation

Deathroels committed:
changed value annotation for kafka topic and add userContext for kafka template key

Deathroels committed:
try to run post view analytics event changes

Deathroels committed:
-change kafka properties in application.yaml

Deathroels committed:
-change Serializer to StringSerializer

Deathroels committed:
-hot changes for correct event working

Deathroels committed:
-remove unused mapper

Deathroels committed:
-add custom task executor, for post events

Deathroels committed:
-fix problem with qualifier annotation

Deathroels committed:
Merge remote-tracking branch 'origin/feature/BJS2-97260-post-view-analytics' into feature/BJS2-97260-post-view-analytics

# Conflicts:
#	src/main/java/faang/school/postservice/dto/post/PostEventAspect.java

Deathroels committed:
Merge branch 'cerberus-hippogriff-master-stream12' into feature/BJS2-97260-post-view-analytics

# Conflicts:
#	src/main/java/faang/school/postservice/service/post/PostV2Service.java

Deathroels committed:
Merge branch 'cerberus-hippogriff-master-stream12' into feature/BJS2-97260-post-view-analytics

# Conflicts:
#	src/main/java/faang/school/postservice/service/post/PostV2Service.java

Deathroels committed:
-del unused import

Deathroels committed:
-fix application.yaml kafka properties

Deathroels committed:
-rename to PostViewEventProducer
-add in KafkaProducerConfig stringKafkaTemplate, and commentEventKafkaTemplate

Deathroels committed:
-add qualifier for kafka configs
-change in Post entity, List to Set
-add primary KafkaTemplate, and others in KafkaProducerConfig

Scherbi1 commented on src/main/java/faang/school/postservice/config/AsyncConfig.java:
```diff
@@ -0,0 +1,47 @@
+        executor.setKeepAliveSeconds(keepAliveSeconds);
+        executor.setRejectedExecutionHandler(new ThreadPoolExecutor.CallerRunsPolicy());
+        executor.setWaitForTasksToCompleteOnShutdown(true);
+        executor.setAwaitTerminationSeconds(30);
```
M-Magic
а почему и это не вынес в конфиг?

Scherbi1 commented on src/main/java/faang/school/postservice/config/KafkaProducerConfig.java:
```diff
@@ -0,0 +1,71 @@
+        props.put(ProducerConfig.VALUE_SERIALIZER_CLASS_CONFIG, JsonSerializer.class);
+
+        return new DefaultKafkaProducerFactory<>(props);
+    }
```
с 27 по 33 строки такое же код или мне кажется?

mipkasnow replied:
да, тут точно можно оптимизировать, например в юзер сервисе можно глнуть. Общий конфиг можно вынести, чтобы новые топики моно было создавать с минимальным дублированием кода

Scherbi1 commented on src/main/java/faang/school/postservice/controller/facade/post/PostFacade.java:
```diff
@@ -42,6 +48,15 @@ public void deleteById(Long postId) {
+                postId,
+                result.getAuthorId(),
+                userContext.getUserId(),
+                LocalDateTime.now()
```
ага
то есть пользователю будешь отправлять id? почему не имена? откуда пользователь знает, что это за ид

Scherbi1 commented on src/main/java/faang/school/postservice/dto/post/PostViewEvent.java:
```diff
@@ -0,0 +1,15 @@
+
+public record PostViewEvent(
+        Long postId,
+        Long authorId,
```
.

20cognosce replied:
одного authorId недостаточно на стороне аналитики, не надо, чтобы аналитика слала get запрос по автору, лучше сразу в ДТОшке пришли все данные

Scherbi1 commented on src/main/java/faang/school/postservice/model/Post.java:
```diff
@@ -51,11 +53,11 @@ public class Post {
     @OneToMany(mappedBy = "post", orphanRemoval = true)
     @Builder.Default
-    private List<Comment> comments = new ArrayList<>();
+    private Set<Comment> comments = new HashSet<>();
```
а почему сет? комментарии не могут повторяться?

Deathroels committed:
-add await termination seconds in application.yaml

Deathroels committed:
-add await termination seconds in application.yaml

Deathroels committed:
Merge remote-tracking branch 'origin/feature/BJS2-97260-post-view-analytics' into feature/BJS2-97260-post-view-analytics

Scherbi1 commented:
С нерабочими тестами не поставлю апрув

Scherbi1 commented on src/main/java/faang/school/postservice/dto/post/PostEventAspect.java:
```diff
@@ -0,0 +1,94 @@
+                    );
+                    eventPublisher.publish(event);
+                }
+            });
```
ой
сложна
а если декомпозировать?

Scherbi1 commented on src/main/java/faang/school/postservice/producer/PostViewEventProducer.java:
```diff
@@ -0,0 +1,32 @@
+    private final KafkaTemplate<String, PostViewEvent> kafkaTemplate;
+
+    public void publish(PostViewEvent event) {
+        String key = event.postId().toString() + event.authorId() + event.currentTime().toString();
```
я не очень люблю сложение строк
хоть и стринг билдер под капотом юзает 
но тригерит

Deathroels committed:
-refactor KafkaProducerConfig

Deathroels committed:
refactor:
-change PostViewEvent Long authorId to UserDto author
-changed similar methods in aspect and changed to the general method
-Removed repetitive functionality
-add readOnly param for @Transactional
-refactor KafkaProducerConfig
-remove standart properties for custom executor pool in application.yaml
-del useless log in PostViewEventProducer
-rollback and fix test

Deathroels committed:
-refactor review changes

Scherbi1 commented:
<img width="400" height="400" alt="image" src="https://github.com/user-attachments/assets/1fdeb9c0-45b8-4470-a144-2699a124ff23" />

Scherbi1 commented on src/main/java/faang/school/postservice/config/AsyncConfig.java:
```diff
@@ -0,0 +1,50 @@
+        executor.setMaxPoolSize(maxSize);
+        executor.setQueueCapacity(queueCapacity);
+        executor.setThreadNamePrefix(threadNamePrefix);
+        executor.setKeepAliveSeconds(keepAliveSeconds);
+        executor.setRejectedExecutionHandler(new ThreadPoolExecutor.CallerRunsPolicy());
+        executor.setWaitForTasksToCompleteOnShutdown(true);
+        executor.setAwaitTerminationSeconds(awaitTerminationSeconds);
+        executor.initialize();
```
то есть прям столько много надо?

Scherbi1 commented on src/main/java/faang/school/postservice/config/KafkaProducerConfig.java:
```diff
@@ -0,0 +1,72 @@
+    @Primary
+    public KafkaTemplate<String, Object> kafkaTemplate() {
+        return createKafkaTemplate(Object.class);
+    }
```
у тебя же есть KafkaTemplate<String, T> и   KafkaTemplate<String, Object>
и использовал @Primary
почему бы имя не задать
вдруг другому нужно будет KafkaTemplate<String, T>
может одного будет достаточно?

Deathroels replied:
убрал эту затею, корректно думаю оставить каждому свой kafkaTemplate пока что, producerFactory на общий

Scherbi1 commented on src/main/java/faang/school/postservice/producer/PostViewEventProducer.java:
```diff
@@ -0,0 +1,32 @@
+    private final KafkaTemplate<String, PostViewEvent> kafkaTemplate;
+
+    public void publish(PostViewEvent event) {
+        String key = String.format("%d%d%s", event.postId(), event.authorId(), event.currentTime());
```
ООООО

Scherbi1 replied:
приятно <3

Scherbi1 commented on src/main/java/faang/school/postservice/producer/PostViewEventProducer.java:
```diff
@@ -0,0 +1,32 @@
+            log.error("Error sending PostViewEvent: {}", e.getMessage());
+        }
+
+        log.debug("PostViewEvent sent asynchronously: {}", event);
```
зачем это?

20cognosce replied:
можно убрать бесполезный лог

20cognosce commented:
Почему 41 коммит

20cognosce commented on src/test/java/faang/school/postservice/service/post/PostV2ServiceTest.java:
```diff
@@ -210,22 +206,6 @@ void deletePostSoftly_shouldDeletePost_whenAllDataValid() {
     }
 
-    @Test
-    void findById_shouldReturnPostDto_whenPostExists() {
```
А почему ты удалил тест...

Deathroels replied:
май бэд!

20cognosce commented on src/main/java/faang/school/postservice/service/post/PostV2Service.java:
```diff
@@ -108,10 +107,13 @@ public PageResponse<PostV2Dto> findAllDraftsByAuthor(Pageable pageable) {
     }
 
+    @PublishPostEvent(eventClass = PostV2Dto.class)
+    @Transactional
```
readOnly=true стоит добавить

20cognosce commented on src/main/java/faang/school/postservice/service/post/PostV2Service.java:
```diff
@@ -90,6 +88,7 @@ public void deletePostSoftly(Long postId) {
     }
 
+    @PublishPostEvent(eventClass = PageResponse.class)
     public PageResponse<PostV2Dto> findAllPublishedByFilter(Long authorId, Pageable pageable) {
```
Сюда тоже повесить транзакцию

20cognosce commented on src/main/java/faang/school/postservice/controller/facade/post/PostFacade.java:
```diff
@@ -42,6 +48,15 @@ public void deleteById(Long postId) {
+                userContext.getUserId(),
+                LocalDateTime.now()
+        );
+        eventPublisher.publish(event);
```
Зачем второй раз?

20cognosce commented on src/main/java/faang/school/postservice/config/KafkaProducerConfig.java:
```diff
@@ -0,0 +1,72 @@
+
+    private <T> KafkaTemplate<String, T> createKafkaTemplate(Class<T> valueType) {
+        return new KafkaTemplate<>(createProducerFactory(valueType));
+    }
```
Приватные методы - внизу
Публичные - вверху

у тебя наоборот

20cognosce commented on src/main/java/faang/school/postservice/config/KafkaProducerConfig.java:
```diff
@@ -0,0 +1,72 @@
+    }
+
+    @Bean
+    @Primary
```
зачем @Primary, с кем конфликтуешь? лучше имя дай бинам, и везде явно указывай квалифаеры
producerFactory в принципе можно одну оставить на весь сервис, и её будет разные кафкаТеплейты юзать

20cognosce commented on src/main/java/faang/school/postservice/config/KafkaProducerConfig.java:
```diff
@@ -0,0 +1,72 @@
+    @Bean
+    @Primary
+    public ProducerFactory<String, Object> producerFactory() {
+        return createProducerFactory(Object.class);
```
Object.class - это костыль конечно

20cognosce commented on src/main/java/faang/school/postservice/dto/post/PostEventAspect.java:
```diff
@@ -0,0 +1,111 @@
+
+    private Collection<PostV2Dto> extractFromCollection(Collection<?> collection) {
+        return extractPostDtosFromSource(collection);
+    }
```
так, а зачем было два метода плодить, если в родительском методе можно было просто вызвать pageResponse.content(), и в этом вся разница

20cognosce reviewed (changes requested):
Хорошо, немного поправить

Deathroels committed:
-rollback and fix test

Deathroels committed:
Merge remote-tracking branch 'origin/feature/BJS2-97260-post-view-analytics' into feature/BJS2-97260-post-view-analytics

# Conflicts:
#	src/main/java/faang/school/postservice/config/KafkaProducerConfig.java
#	src/main/java/faang/school/postservice/dto/post/PostEventAspect.java
#	src/main/java/faang/school/postservice/producer/PostViewEventProducer.java

Deathroels commented:
Ковырял, ковырял, да пришлось асинхронность для аспекта выключить как промежуточный костыль
выдавал ошибку
"
22:35:16.699 [PostEvent-1] ERROR f.s.p.dto.post.PostEventAspect - Failed to publish PostViewEvent: User ID not found in context
"
Но меня пока не хватило на ее решение

20cognosce commented on src/main/java/faang/school/postservice/dto/post/PostViewEvent.java:
```diff
@@ -0,0 +1,16 @@
+        Long postId,
+        UserDto author,
+        Long viewerId,
+        @JsonFormat(pattern = "yyyy-MM-dd HH:mm:ss")
```
Это не стандартный формат написания даты

20cognosce commented on src/main/java/faang/school/postservice/producer/PostViewEventProducer.java:
```diff
@@ -0,0 +1,33 @@
+    @Value("${kafka.topic.post-view}")
+    private String topicName;
+
+    @Qualifier("postViewEventKafkaTemplate")
```
вместе с ломбоком был баг, что не работает
нужно заглянуть в .class и посмотреть, научился ли ломбок читать внешние аннотации при генерации конструктора
